dash/_hooks.py

```diff
@@ -0,0 +1,93 @@
+import typing as _t
+
+_ns = {
+    "setup": [],
+    "layout": [],
+    "routes": [],
+    "error": [],
+    "callback": [],
+}
+
+
+def layout(func):
+    """
+    Run a function when serving the layout, the return value
+    will be used as the layout.
+    """
+    _ns["layout"].append(func)
+    return func
+
+
+def setup(func):
+    """
+    Can be used to get a reference to the app after it is instantiated.
+    """
+    _ns["setup"].append(func)
+    return func
+
+
+def route(name=None, methods=("GET",)):
+    """
+    Add a route to the Dash server.
+    """
+
+    def wrap(func):
+        _name = name or func.__name__
+        _ns["routes"].append((_name, func, methods))
+        return func
+
+    return wrap
+
+
+def error(func: _t.Callable[[Exception], _t.Any]):
+    """Automatically add an error handler to the dash app."""
+    _ns["error"].append(func)
+    return func
+
+
+def callback(*args, **kwargs):
+    """
+    Add a callback to all the apps with the hook installed.
+    """
+
+    def wrap(func):
+        _ns["callback"].append((list(args), dict(kwargs), func))
+        return func
+
+    return wrap
+
+
+class HooksManager:
+    _registered = False
```

Comment:
please add a comment explaining what `_registered` is for - in particular, why is it a class-level variable instead of an instance variable?

Reply:
see above